src/org/sosy_lab/java_smt/solvers/z3/Z3AbstractProver.java

```diff
@@ -208,27 +215,31 @@ protected long addConstraint0(BooleanFormula f) throws InterruptedException {
     return e;
   }
 
-  @Override
-  public void push() throws InterruptedException {
+  protected void push0() throws InterruptedException {
     Preconditions.checkState(!closed);
-    try {
-      Native.solverPush(z3context, z3solver);
-    } catch (Z3Exception exception) {
-      throw creator.handleZ3Exception(exception);
+    super.push();
+    if (storedConstraints != null) {
+      storedConstraints.push(storedConstraints.peek());
```

Comment:
This puts the previous level on the new level, is this on purpose?

Reply:
This is on purpose, because:
- we use the same strategy in several places (all places are part of this PR).
- we always require the complete union of all constraints when it comes to unsat-cores or interpolants.
- the PersistantMap is quite memory efficient and shares common sub-mappings in most cases.